src/passes/OptimizeInstructions.cpp

```diff
@@ -1804,6 +1823,43 @@ struct OptimizeInstructions
       // Don't bother when `ifFalse` isn't pure - we would need to reverse the
       // order using a temp local, which would be bad
     }
+    {
+      // TODO: Remove this after landing SCCP pass. See: #4161
+
+      // i32(x) ? i32(x) : 0  ==>  x
+      Expression *x, *y;
+      if (matches(curr, select(any(&x), i32(0), any(&y))) &&
+          areConsecutiveInputsEqualAndFoldable(x, y)) {
+        return curr->ifTrue;
+      }
+      // i32(x) ? 0 : i32(x)  ==>  { x, 0 }
+      if (matches(curr, select(i32(0), any(&x), any(&y))) &&
+          areConsecutiveInputsEqualAndFoldable(x, y)) {
+        return builder.makeSequence(builder.makeDrop(x), curr->ifTrue);
+      }
+
+      // i64(x) == 0 ? 0 : i64(x)  ==>  x
+      // i64(x) != 0 ? i64(x) : 0  ==>  x
```

Comment:
I believe we don't need some of these rules, as we can depend on other patterns to simplify them. Specifically, we don't need to match this:
```
i64(x) != 0 ? i64(x) : 0
```
That will be turned into this anyhow:
```
i64(x) ? 0 : i64(x)
```
That is, other patterns will flip the condition and flip the arms of the select. And they can do that here because the arms can be reordered - one of them is a constant, so that is always true.

Reply:
I don't understand how we could do this: `i64(x) ? 0 : i64(x)`. `i64` is not valid for truish value as argument for select/if. So we need course i64 to i32 anyway.

Reply:
Oh, right, thanks, I forgot that an i64 is not a valid input here...